### PR TITLE
hail::hstring and hstringstream avoid std::string ABI incompatibility

### PR DESCRIPTION
**Pull request by rcownie**
The default OS for dataproc instances is based on debian8, which uses g++-4.9.x
That has a libstdc++ with an old-ABI implementation of std::list and std::string.  To build
a libhail.so which can link against the default libstdc++ on g++-4.x systems, we need to
avoid the use of std::string inside libhail.so (but it's ok to use it in dynamic-generated code,
which will be built with a compiler which matches the libstdc++).

This commit introduces a minimal hail::hstring and hail::hstringstream with the necessary
functionality for NativeModule.cpp.

Since these don't have a std::hash, I also imported the source code for the (free and uncopyrighted)
MurmurHash3, a fast high-quality (but non-crypto) hash function which can give a 128bit hash.  
This simplifies the calculation of the 80bit hash used for module-keys.

In addition to using these prebuilt libraries, we also need to get g++ installed on the dataproc
master node, which could be done with "sudo apt-get install build-essential".  But I'm not yet sure where
that initialization step needs to go.

### CODE REVIEW

**Comment by danking** [comment]
I rolled the PR dice and assigned amanda

**Comment by rcownie** [comment]
Confirmed that this prebuilt libhail.so can run tests with HAIL_ENABLE_CPP_CODEGEN=1
on a dataproc node with the default 1.2 image (debian8 and g++-4.9.2).

The dataproc nodes have g++ and make already installed, so this PR should suffice to
make them work with C++ codegen.

**Comment by cseed** [comment]
So I'm a little confused why this hailstring stuff is back.  I thought the plan was to ship the compiler used to build libhail.so with Hail so there is no possibility of incompatibility.

**Comment by rcownie** [comment]
1. It's feasible to build and ship the compiler + libraries for a limited number of known platforms
    (at Physics Speed I did this for Ubuntu-16.04 and one particular version of CentOS).  It gets nuts if
   you have many different OS'es each of which needs its own compiler build (and it then becomes
   another build-system/packaging issue to get all those compilers built correctly for each OS).
   Possibly a good thing to do in the long run.  Probably not something I could do in the limited time
   available.

2. If you build your own compiler + library, then you risk becoming incompatible with other 
   libraries on the target system which were built against that system's "standard" compiler
  and library and header files.  e.g. BLAS.  [Though this only applies to libraries compiled from C++,
  not libraries in C, which might conatin the damage].

So it seemed like the least disruptive path in the short term was to excise the few uses of
std::string and std::stringstream, so that we can build a libhail.so which should work across
a wide variety of Linux systems.

**Comment by cseed** [comment]
There is no requirement to support many OSes.  We only have to support two platforms: recent Linux and OSX.  They have variants but should be mutually compatible.  Testing on Dataproc + recent version of Ubuntu or Debian (which we do in the CI) seems fine.  OSX has flags for version support, we can just pick a version (10.10, say) and build for that and beyond.  As I've said before, we control what we support and there is no need to make this a burden on ourselves.

 > If you build your own compiler + library, then you risk becoming incompatible with other libraries

Yes, we should ship with all our C++ dependencies.  You convinced me of this?  Then there is no issue.  BLAS is a C library, so no issue there.  Right now I think that just means the compiler and the standard library.

 > Probably not something I could do in the limited time available.

Fair.  I'm happy with partial progress in in the above direction, but this seems like a step backwards and something we will want to revert soon.  I'm not inclined to go in this direction.  Not having access to the standard library seems problematic.

 > Confirmed that this prebuilt libhail.so can run tests with HAIL_ENABLE_CPP_CODEGEN=1 on a dataproc node with the default 1.2 image (debian8 and g++-4.9.2).

Did you test submitting jobs to the cluster itself?  This can be quite a different environment than the tests.

Also, is there a plan about how users (or we) control this in the Dataproc setting?  E.g. how do we submit cluster_sanity_check.py with and without C++ codegen enabled?

**Comment by rcownie** [comment]
>Not having access to the standard library seems problematic.

You absolutely have access to the full C++11 standard library in dynamically-generated code -
you're compiling with the master node's default compiler, and header files, and using
the default libstdc++.o (libc++.dylib), and it's all fine.  And you'll be using whichever flavor
of std::string is the default for that system, which will presumably also be interoperable with any
third-party library packages on that system.

The issues we're getting round are:

a) If libhail.so is prebuilt on a new-ABI system *and* uses std::string, then it can't run against
    the libstdc++ on an old-ABI system

b) If libhail.so is prebuilt on an old-ABI system, then it can run against either old-ABI or new-ABI
     libstdc++, but if it then gets linked against third-party libraries compiled against new-ABI
     headers, you'll have two different flavors of std::string floating around in the same program,
     which causes confusion at any interfaceswhich pass std::string around.

Now if you want to go further in shipping more of the system, then the question of whether to
ship your own libstdc++ (or libc++) is independent of the choice of compiler version.  *If* you 
ship your own libstdc++, then you potentially introduce the problems of interoperability with
other libraries on the system).   And there's a slight question about whether your libstdc++ will
work against the other systems libc.so, though I think the ABI at that level has been stable enough for long enough that it probably doesn't cause trouble.
                       
In short, it's a can of worms.  Avoiding std::string in libhail.so keeps the can closed for now.
And I believe dataproc will move to using debian9 images as the default in November, so at
some point the need to support old-ABI systems will diminish and possibly go away completely.

**Comment by cseed** [comment]
> And there's a slight question about whether your libstdc++ will work against the other systems libc.so

Is there something about libstdc++ here?  Our code will certainly call libc directly, too. As far as I know there no proposal on the table handles incompatible libc's.  I think we'd have to make multiple builds.  Luckily, as far as I know, recent distributions have compatible libc so this shouldn't be an issue.

 > Avoiding std::string in libhail.so keeps the can closed for now.

Can you explain why std::string is special here vs the rest of the standard library?  What assumption are you working under?

**Comment by rcownie** [comment]
>Can you explain why std::string is special here vs the rest of the standard library? What assumption >are you working under?

Yes. The C++11 final standard introduced some new constraints on the definition of std::string
and std::list (essentially outlawing a copy-on-write implementation of std::string, and requiring
that std::list::size() must be O(1)).

Consequently libstdc++ had to be redesigned with incompatible implementations of 
std::string and std::list to become C++11-compliant.  Systems using the earlier libstdc++
(which also typically have g++-4.8.x/4.9x) have the old-ABI not-fully-compliant std::string
(and std::list, but I've gone 20 years without ever using that).

Later versions of libstdc++ have *both* flavors of std::string, but use namespaces to allow them
to coexist (but not to be interchangeable, so interfaces between old-ABI and new-ABI are
problematic).

https://gcc.gnu.org/onlinedocs/libstdc++/manual/using_dual_abi.html

"In the GCC 5.1 release libstdc++ introduced a new library ABI that includes new implementations of std::string and std::list. These changes were necessary to conform to the 2011 C++ standard which forbids Copy-On-Write strings and requires lists to keep track of their size.

In order to maintain backwards compatibility for existing code linked to libstdc++ the library's soname has not changed and the old implementations are still supported in parallel with the new ones. This is achieved by defining the new implementations in an inline namespace so they have different names for linkage purposes, e.g. the new version of std::list<int> is actually defined as std::__cxx11::list<int>. Because the symbols for the new implementations have different names the definitions for both versions can be present in the same library."

**Comment by rcownie** [comment]
https://writingexplained.org/idiom-dictionary/you-can-lead-a-horse-to-water
